javalib/src/main/scala/java/net/URI.scala

```diff
@@ -601,39 +614,99 @@ object URI {
 
   // Quote helpers
 
-  private val quoteChar: js.Function1[String, String] = { (str: String) =>
-    require(str.length == 1)
+  private def decodeComponent(str: String): String = {
+    // Fast-track, if no encoded components
+    if (str.forall(_ != '%')) str
+    else {
+      val inBuf = CharBuffer.wrap(str)
+      val outBuf = CharBuffer.allocate(inBuf.capacity)
+      val byteBuf = ByteBuffer.allocate(64)
+      val decoder = StandardCharsets.UTF_8.newDecoder
+        .onMalformedInput(CodingErrorAction.REPLACE)
+        .onUnmappableCharacter(CodingErrorAction.REPLACE)
+
+      def decode(endOfInput: Boolean) = {
+        byteBuf.flip()
+        decoder.decode(byteBuf, outBuf, endOfInput)
+        if (endOfInput) {
+          decoder.reset()
+          byteBuf.clear()
+        } else {
+          byteBuf.compact()
+        }
+      }
+
+      def flush() = {
+        if (byteBuf.position > 0) decode(true)
```

Comment:
This condition is too restrictive. `decode(..., endOfInput = true)` is supposed to be called at the end of all decoding operations, in case the decoder does buffering. An explicit boolean flag would fix this.

And while we're here, `decode(true)` should be on its own line, since it is a side-effecting `then` part.